blockchain/blocklocator.go

```diff
@@ -6,139 +6,94 @@ package blockchain
-		}
-
-		blockHeight = height
+	// Calculate the max number of entries that will ultimately be in the
+	// block locator.  See the description of the algorithm for how these
+	// numbers are derived.
+	var maxEntries uint8
+	if node.height <= 12 {
+		maxEntries = uint8(node.height) + 1
 	} else {
-		blockHeight = node.height
-
-		// Find the height at which this node forks from the main chain
-		// if the node is on a side chain.
-		if !node.inMainChain {
-			for n := node; n.parent != nil; n = n.parent {
-				if n.inMainChain {
-					forkHeight = n.height
-					break
-				}
-			}
-		}
+		// Requested hash itself + previous 10 entries + genesis block.
+		// Then floor(log2(height-10)) entries for the skip portion.
+		adjustedHeight := uint32(node.height) - 10
+		maxEntries = 12 + fastLog2Floor(adjustedHeight)
 	}
+	locator := make(BlockLocator, 0, maxEntries)
 
-	// Generate the block locators according to the algorithm described in
-	// in the BlockLocator comment and make sure to leave room for the final
-	// genesis hash.
-	//
-	// The error is intentionally ignored here since the only way the code
-	// could fail is if there is something wrong with the database which
-	// will be caught in short order anyways and it's also safe to ignore
-	// block locators.
-	_ = bi.db.View(func(dbTx database.Tx) error {
-		iterNode := node
-		increment := int32(1)
-		for len(locator) < wire.MaxBlockLocatorsPerMsg-1 {
-			// Once there are 10 locators, exponentially increase
-			// the distance between each block locator.
-			if len(locator) > 10 {
-				increment *= 2
-			}
-			blockHeight -= increment
-			if blockHeight < 1 {
-				break
-			}
+	step := int32(1)
+	for node != nil {
+		locator = append(locator, &node.hash)
 
-			// As long as this is still on the side chain, walk
-			// backwards along the side chain nodes to each block
-			// height.
-			if forkHeight != -1 && blockHeight > forkHeight {
-				for iterNode != nil && blockHeight > iterNode.height {
-					iterNode = iterNode.parent
-				}
-				if iterNode != nil && iterNode.height == blockHeight {
-					locator = append(locator, &iterNode.hash)
-				}
-				continue
-			}
+		// Nothing more to add once the genesis block has been added.
+		if node.height == 0 {
+			break
+		}
 
-			// The desired block height is in the main chain, so
-			// look it up from the main chain database.
-			h, err := dbFetchHashByHeight(dbTx, blockHeight)
-			if err != nil {
-				// This shouldn't happen and it's ok to ignore
-				// block locators, so just continue to the next
-				// one.
-				log.Warnf("Lookup of known valid height failed %v",
-					blockHeight)
-				continue
-			}
-			locator = append(locator, h)
+		// Calculate height of previous node to include ensuring the
+		// final node is the genesis block.
+		height := node.height - step
+		if height < 0 {
+			height = 0
 		}
 
-		return nil
-	})
+		// Walk backwards through the nodes to the correct ancestor.
+		node = node.Ancestor(height)
```

Comment:
This suggestion gets implemented on the chain view in the next PR #1013 which does take advantage of the O(1) lookups they provide.  I could certainly do an index lookup here too, but I figured since #1014 is already done (switch over to using the chain view) and #1013 has the optimization, I'd just let it go.

Great observation!